src/Cabal2JSON.hs

```diff
@@ -188,25 +260,79 @@ instance HasCodec SourceRepo where
        <*> requiredField' "tag" .= repoTag
        <*> requiredField' "subdir" .= repoSubdir
 
-instance HasCodec VersionRange where
-  -- TODO: ?
-  codec = undefined
-
 instance HasCodec ShortText where
-  -- TODO: ?
-  codec = undefined
+  codec = dimapCodec toShortText fromShortText codec
 
 instance HasCodec Cabal.License where
-  -- TODO: ?
-  codec = undefined
+  -- TODO add others
+  codec = stringConstCodec
+    [ (BSD2, "BSD2")
+    , (BSD3, "BSD3")
+    , (BSD4, "BSD4")
+    , (MIT, "MIT")
+    , (ISC, "ISC")
+    , (PublicDomain, "PublicDomain")
+    , (AllRightsReserved, "AllRightsReserved")
+    , (UnspecifiedLicense, "UnspecifiedLicense")
+    , (OtherLicense, "OtherLicense")
+    ]
 
 instance HasCodec SPDX.License where
-  -- TODO: ?
-  codec = undefined
+  codec = object "spdx-license" $ dimapCodec f g $
+      eitherCodec (pure NONE) $
+        requiredField' "license-expression"
+    where
+      f = \case
+        Left _ -> NONE
+        Right exp -> License exp
+      g = \case
+        NONE -> Left "None"
+        License exp -> Right exp
+
+instance HasCodec SPDX.LicenseExpression where
+  codec = object "LicenseExpression" $ dimapCodec f g $
+    eitherCodec ((,) <$> requiredField' "expression" .= fst <*> requiredField' "exception-id" .= snd) $
+      eitherCodec ((,) <$> requiredField' "expression-1" .= fst <*> requiredField' "expression-2" .= snd) $
+        (,) <$> requiredField' "expression-1" .= fst <*> requiredField' "expression-2" .= snd
```

Comment:
This one is currently not working. I think its because its not disjoint right? How do I deal with this?

Reply:
I have tried giving each field individual names but its still giving me `thread blocked indefinitely in an MVar operation`

```hs
instance HasCodec SPDX.LicenseExpression where
  codec = dimapCodec f g $
    eitherCodec (object "ELicense" $ (,) <$> requiredField' "expression" .= fst <*> requiredField' "exception-id" .= snd) $
      eitherCodec (object "EAnd" $ (,) <$> requiredField' "expression-1" .= fst <*> requiredField' "expression-2" .= snd) $
        object "EOr" $ (,) <$> requiredField' "expression-11" .= fst <*> requiredField' "expression-22" .= snd
    where
      f = \case
        Left (exp, exc) -> ELicense exp exc
        Right (Left (exp1, exp2)) -> EAnd exp1 exp2
        Right (Right (exp1, exp2)) -> EOr exp1 exp2
      g = \case
        ELicense exp exc -> Left (exp, exc)
        EAnd exp1 exp2 -> Right $ Left(exp1, exp2)
        EOr exp1 exp2 -> Right $ Right (exp1, exp2)
```

@NorfairKing

Reply:
@m1-s I haven't gotten to it. I think @marijanp might be able to respond on shorter notice but I'll do my best to have a look shortly..

Reply:
How do you know it's not working?
I'm getting the following test failure:
```
Failures:

    test/Cabal2JSONSpec.hs:48
  ✗ 1 Cabal2JSONSpec.outputs the same json for the cabal2json cabal file
      Error in $: not enough input
```

Reply:
yeah i might have forgotten to commit that one...

Reply:
> How do you know it's not working? I'm getting the following test failure:
> 
> ```
> Failures:
> 
>     test/Cabal2JSONSpec.hs:48
>   ✗ 1 Cabal2JSONSpec.outputs the same json for the cabal2json cabal file
>       Error in $: not enough input
> ```

Thats the wrong test. Uncomment the `SPDX.LicenseExpression` schemaSpec test to see it fail (or checkout latest commit that i have just updated). I had that test disabled because it crashes the whole test suite. 

Execute `stack test` and you should get this output:

```
  ✓ outputs the same yaml file for benchmark-type                          0.13 ms
  ✓ outputs the same yaml file for foreign-lib-option                      0.14 ms
  ✓ outputs the same yaml file for module-reexport                         0.13 ms
cabal2json-test: thread blocked indefinitely in an MVar operation
```

There is actually no error message that indicates which test is blocking but if you comment out the `@SPDX.LicenseExpression` test, the test suite works, so it has to be that one.